File: Data Structures & Algorithms/path-with-minimum-effort/submission-3.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        # Time Complexity: O(rows * cols * log(rows * cols)) Space Complexity: O(rows * cols)
        minHeap = []
        visited = set()
        moves = ((0,1), (1,0), (0,-1,), (-1,0))
        heapq.heappush(minHeap, [0, 0, 0]) # cost, row, col

        while minHeap:
            cost, r, c = heapq.heappop(minHeap)
            if (r, c) in visited: continue
            visited.add((r, c))
            if r == len(heights) - 1 and c == len(heights[0]) - 1: return cost

            for dr, dc in moves:
                nr, nc = r + dr, c + dc
                if min(nr, nc) >= 0 and nr < len(heights) and nc < len(heights[nr]) and (nr, nc) not in visited:
                    ncost = abs(heights[r][c] - heights[nr][nc])
                    heapq.heappush(minHeap, [max(cost, ncost), nr, nc])

File: Data Structures & Algorithms/max-water-container/submission-3.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        // Time Complexity: O(n) Space Complexity: O(1)
        int l{0}, r{int(heights.size() - 1)};
        int ans{0};
        while (l < r) {
            ans = max(ans,min(heights[l], heights[r]) * (r - l));
            if (heights[l] > heights[r]) r--;
            else l++;
        }
        return ans;
    }
};
